packages/lit-ssr/src/lib/render-lit-html.ts

```diff
@@ -0,0 +1,551 @@
+            ops.push({
+              type: 'custom-element-open',
+              tagName,
+              ctor,
+              staticAttributes: new Map(
+                node.attrs
+                  .filter((attr) => !attr.name.endsWith(boundAttributeSuffix))
+                  .map((attr) => [attr.name, attr.value])
+              ),
+            });
+          }
+        }
+        if (node.attrs.length > 0) {
+          for (const attr of node.attrs) {
+            if (attr.name.endsWith(boundAttributeSuffix)) {
+              writeTag = true;
+              boundAttrsCount += 1;
+              // Note that although we emit a lit-bindings comment marker for any
+              // nodes with bindings, we don't account for it in the nodeIndex because
+              // that will not be injected into the client template
+              const name = attr.name.substring(
+                0,
+                attr.name.length - boundAttributeSuffix.length
+              );
+              const strings = attr.value.split(marker);
+              const attrSourceLocation = node.sourceCodeLocation!.attrs[
+                attr.name
+              ];
+              const attrNameStartOffset = attrSourceLocation.startOffset;
+              const attrEndOffset = attrSourceLocation.endOffset;
+              flushTo(attrNameStartOffset);
+              ops.push({
+                type: 'attribute-part',
+                index: nodeIndex,
+                name,
+                strings,
+                tagName,
+                useCustomElementInstance: ctor !== undefined,
+              });
+              skipTo(attrEndOffset);
+            } else if (node.isDefinedCustomElement) {
+              // We will wait until after connectedCallback() and render all
+              // custom element attributes then
+              const attrSourceLocation = node.sourceCodeLocation!.attrs[
+                attr.name
+              ];
+              flushTo(attrSourceLocation.startOffset);
+              skipTo(attrSourceLocation.endOffset);
+            }
+          }
+        }
+
+        if (writeTag) {
+          if (node.isDefinedCustomElement) {
+            flushTo(node.sourceCodeLocation!.startTag.endOffset - 1);
+            ops.push({
+              type: 'custom-element-attributes',
+            });
+            flush('>');
+            skipTo(node.sourceCodeLocation!.startTag.endOffset);
+          } else {
+            flushTo(node.sourceCodeLocation!.startTag.endOffset);
+          }
+        }
+
+        if (boundAttrsCount > 0) {
+          flush(`<!--lit-bindings ${nodeIndex}-->`);
+        }
+
+        if (ctor !== undefined) {
+          ops.push({
+            type: 'custom-element-children',
+          });
+        }
+      }
+      nodeIndex++;
+    },
+    post(node) {
+      if (isElement(node) && node.isDefinedCustomElement) {
+        ops.push({
+          type: 'custom-element-close',
+        });
+      }
+    },
+  });
+  flushTo();
```

Comment:
What does this do?

Reply:
Flushes to the end of the template

Reply:
Added comment. When no parameter is given, it flushes any remaining text. I think the only remaining text would be e.g. close tags, since any top-level text would constitute a text node, which would fully flush the remaining text.

Reply:
ok